Add spans (#3)
catawiki/opentelemetry-ruby-contrib: instrumentation/grape/README.md

# OpenTelemetry Grape Instrumentation

The Grape instrumentation is a community-maintained instrumentation for [Grape](https://github.com/ruby-grape/grape), a REST-like API framework for Ruby.

> It relies mostly on the Grape built-in support for `ActiveSupport::Notifications` (more info [here](https://github.com/ruby-grape/grape#active-support-instrumentation)).

Reviewer: ```suggestion
It relies on the Grape built-in support for `ActiveSupport::Notifications` (more info [here](https://github.com/ruby-grape/grape#active-support-instrumentation)).
```

Author: done ✅

It currently supports the following events:

- `endpoint_run.grape`
- `endpoint_render.grape`
- `endpoint_run_filters.grape`

> - `format_response.grape`

Author: I implemented this in a separate commit because I don't know if we will keep it 🤔 
The good: it's an extra event which, as @dhruvCW said, can fail due to issues with entities and such.
The bad: it is not a child of `endpoint_run`, so we'll be relying on Rack to connect it to the trace. I'm fine with this but I don't know if the maintainers will be, but we can see 😄 
Bear in mind that since `endpoint_run` is not a parent, and `format_response` happens later, if `format_response` blows up the `http.status_code` for `endpoint_run` will still be 200 😢

Reviewer 2: >if format_response blows up the http.status_code for endpoint_run will still be 200 😢

true but depending on the defined middleware this response can be an error response and thus tracked by an upstream span (I am thinking rack middleware generated span)

PS: not disagreeing adding to the information presented here.

Author: Thank you for the clarification 🙇‍♀️  And yes, I think it will be ok 😄
But I just wanted to point it out (and I plan to leave this comment in the upstream PR so the maintainers are fully aware).

end
```

> Since Grape is "designed to run on Rack or complement existing web application frameworks such as Rails and Sinatra", we recommend using it along with the Rack, Rails and/or Sinatra instrumentations.

Reviewer: nit:

```suggestion
Grape is "designed to run on Rack or complement existing web application frameworks such as Rails and Sinatra", it is recommended to use this instrumentation along with the Rack, Rails and/or Sinatra instrumentations.
```

Author: done ✅

> 

Reviewer 2: nice ❤️  🚀

> Since Grape is "designed to run on Rack or complement existing web application frameworks such as Rails and Sinatra", it is recommended to use this instrumentation it along with the Rack, Rails and/or Sinatra instrumentations.

Reviewer: ```suggestion
Since Grape is "designed to run on Rack or complement existing web application frameworks such as Rails and Sinatra", it is recommended to use this instrumentation along with the Rack, Rails and/or Sinatra instrumentations.
```

Was probably my bad, sorry!

end
```

### Configuration options

#### `:ignored_events` (array)

Indicate if any events should not produce spans.

> - Accepted values: `:endpoint_render`, `:endpoint_run_filters`.

Reviewer 2: why isn't `format_response` an option ?

TBH `endpoint_run` could also be excluded and then the generated spans are not parented at all (except the root span parent).

Author: Because I forgot to add it 😅 

> TBH endpoint_run could also be excluded and then the generated spans are not parented at all (except the root span parent).

Yes, you're right, but I didn't add it as an option because since it's the "main" span, I thought it wouldn't make a lot of sense to ignore this one, which is also the one who has all the attributes, and be left with all the orphans 😅

- Defaults to `[]` (no ignored events).

Example:

```ruby
OpenTelemetry::SDK.configure do |c|
  c.use 'OpenTelemetry::Instrumentation::Grape', { ignored_events: [:endpoint_run_filters] }
end
```

> Note that the `endpoint_run` event cannot be disabled since it is the parent event. If you need to disable the instrumentation, set `:enabled` to `false`:

Reviewer 2: I don't think I agree with that reasoning.

Reviewer: ```suggestion
Note that the `endpoint_run` is unable to be ignored. If you need to disable the instrumentation, set `:enabled` to `false`:
```

Author: I would like to still exclude `endpoint_run` from the "ignorable" events, if you're ok with that, for the reason mentioned in [this comment](https://github.com/catawiki/opentelemetry-ruby-contrib/pull/3#discussion_r1144347132). It is indeed possible to exclude it, but I think that if we allowed that, we should add a big warning to the README explaining that if you ignore `endpoint_run`, all the other events will now be children of a span produced by another middleware, and I would like to keep it as simple as possible (at least in this first implementation).

I have made the change Ben suggested to the README to remove its being the parent as the cause. Now it just says that it cannot be ignored and that's it 😅
